pkg/kubectl/cmd/delete.go

```diff
@@ -308,20 +329,68 @@ func DeleteResult(r *resource.Result, out io.Writer, ignoreNotFound bool, shortO
 	return nil
 }
 
-func cascadingDeleteResource(info *resource.Info, out io.Writer, shortOutput bool, mapper meta.RESTMapper) error {
+func cascadingDeleteResource(info *resource.Info, out io.Writer, shortOutput bool, mapper meta.RESTMapper, waitForDeletion bool, timeout time.Duration) error {
 	falseVar := false
 	deleteOptions := &metav1.DeleteOptions{OrphanDependents: &falseVar}
-	return deleteResource(info, out, shortOutput, mapper, deleteOptions)
-}
-
-func deleteResource(info *resource.Info, out io.Writer, shortOutput bool, mapper meta.RESTMapper, deleteOptions *metav1.DeleteOptions) error {
-	if err := resource.NewHelper(info.Client, info.Mapping).DeleteWithOptions(info.Namespace, info.Name, deleteOptions); err != nil {
-		return cmdutil.AddSourceToErr("deleting", info.Source, err)
+	obj, err := deleteResource(info, out, shortOutput, mapper, deleteOptions, !waitForDeletion)
+	if err != nil {
+		return err
+	}
+	if !waitForDeletion {
+		return nil
+	}
+	// Keep deleting the resource with uid precondition until we get a
+	// resource not found error.
+	// We use DELETE here instead of using GET to ensure that DELETE
+	// works without requiring GET permissions.
+	uid, err := getUID(obj)
+	if err != nil {
+		return fmt.Errorf("unexpected error in extracting uid: %s", err)
+	}
+	deleteOptions = &metav1.DeleteOptions{OrphanDependents: &falseVar, Preconditions: metav1.NewUIDPreconditions(uid)}
```

Comment:
I assume you want to wait for the dependents to be deleted as well? If so, don't set OrphanDependents, it's deprecated, instead, setting [PropagationPolicy](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go#L401) to [DeletePropagationForeground]`https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go#L370`.